### PR TITLE
Unify CPU and GPU inertia validation semantics

### LINKED ISSUE #2010: Unify CPU and GPU inertia validation semantics
## Summary

Newton currently has two different inertia-validation implementations during
`ModelBuilder.finalize()`:

- Detailed CPU path: `verify_and_correct_inertia()` in
  `newton/_src/geometry/inertia.py`, enabled with
  `validate_inertia_detailed=True`
- Fast/default Warp path: `validate_and_correct_inertia_kernel()` in
  `newton/_src/geometry/inertia.py`, used when
  `validate_inertia_detailed=False`

These two paths do not currently agree on several important behaviors. Some of
the differences are only about warnings, but several are semantic and lead to
different mass / inertia values in the returned `Model`, different
`ModelBuilder` state after `finalize()`, or even a crash in the detailed path.

## Relevant Code Paths

- `newton/_src/sim/builder.py`
  - `ModelBuilder.finalize()`
- `newton/_src/geometry/inertia.py`
  - `verify_and_correct_inertia()`
  - `validate_and_correct_inertia_kernel()`

### 1. `bound_mass` applies differently to zero / negative mass

Detailed CPU path:

- negative mass is set to `0.0`
- `bound_mass` only clamps small positive mass (`mass > 0`)
- zero-mass bodies stay zero-mass

Fast/default Warp path:

- negative mass is set to `0.0`
- `bound_mass` is then applied to any `mass < bound_mass`
- zero-mass and corrected-negative-mass bodies become positive-mass

Example:

- input: `mass=0.0`, `bound_mass=0.1`, `inertia=diag(1,1,1)`
- detailed result: `mass=0.0`, inertia zeroed
- fast result: `mass=0.1`, inertia left as `diag(1,1,1)`

This is a user-visible semantic mismatch in the returned `Model`.

### 2. Positive-mass singular inertia is repaired in CPU mode but left singular in fast mode

For positive-mass bodies with singular inertia, the detailed path makes the
matrix strictly positive definite by adding a small diagonal epsilon. The fast
path can leave singular inertia unchanged.

Examples:

- `mass=1.0`, `inertia=zeros((3,3))`
- `mass=1.0`, `inertia=diag(0,1,1)`

Observed behavior:

- detailed result: adds approximately `1e-6 * I`
- fast result: leaves the singular matrix unchanged

Consequences:

- different corrected inertia values
- different `body_inv_inertia`
- fast path may report no correction in cases the detailed path repairs

### 3. Non-symmetric inertia is symmetrized in CPU mode but preserved in fast mode

The detailed path explicitly enforces symmetry with `(I + I^T) / 2` before
eigenvalue-based validation. The fast path runs `wp.eig3()` directly on the
incoming matrix and can return a non-symmetric inertia tensor unchanged.

Example:

- input:

```python
np.array([
    [1.0, 2.0, 0.0],
    [0.0, 1.0, 0.0],
    [0.0, 0.0, 1.0],
], dtype=np.float32)
```

Observed behavior:

- detailed result: symmetric matrix, then further corrected / balanced
- fast result: non-symmetric matrix preserved in the returned `Model`

This is a serious semantic mismatch because physically valid inertia tensors
should be symmetric.

### 4. Exact triangle-inequality boundary can be a no-op in CPU mode but a correction in fast mode

The detailed path treats exact equality as valid. The fast path currently
corrects some exact-boundary or near-boundary cases because of different
tolerance / eigen decomposition behavior.

Example:

- input: `mass=1.0`, `inertia=diag(1,1,2)`

Observed behavior:

- detailed result: no warning, no correction
- fast result: summary warning and correction to approximately
  `diag(1.0000029, 1.0000029, 2.0000029)`

This means the fast path can emit a false positive correction for a valid
inertia tensor.

### 5. Near-boundary triangle-inequality cases get different correction magnitudes

Even when both paths decide to correct, the final corrected inertia can differ
materially near the boundary because:

- the detailed path uses a `1e-10` triangle-inequality tolerance
- the fast path uses a `1e-6` tolerance
- the eigen decomposition / sorting behavior differs between NumPy and Warp

This is less severe than the exact-boundary false positive above, but it still
breaks parity between the two implementations.

### 6. `NaN` / `Inf` inertia crashes the detailed path but slips through the fast path

This is currently the most severe discrepancy.

Example:

- input: `mass=1.0`, `inertia=diag(np.nan, 1.0, 1.0)`

Observed behavior:

- detailed result: `verify_and_correct_inertia()` raises
  `UnboundLocalError: cannot access local variable 'eigenvalues' where it is not associated with a value`
- fast result: `finalize()` succeeds and returns a `Model` whose inertia still
  contains `NaN`

### 7. `ModelBuilder` state after `finalize()` differs between modes

Detailed mode updates the builder's internal lists in place:

- `builder.body_mass`
- `builder.body_inertia`
- `builder.body_inv_mass`
- `builder.body_inv_inertia`

Fast mode only updates the arrays attached to the returned `Model` and leaves
the builder's internal Python-side state unchanged.

Example:

- input: `mass=-1.0`
- after `finalize(validate_inertia_detailed=True)`: `builder.body_mass[i] == 0.0`
- after `finalize(validate_inertia_detailed=False)`: `builder.body_mass[i]` is
  still `-1.0`, while `model.body_mass[i] == 0.0`

This makes repeated `finalize()` calls and post-finalize builder inspection
mode-dependent.

## Minimal Repro Snippet

```python
import numpy as np
import newton
import warp as wp


def run_case(detailed: bool, mass: float, inertia: np.ndarray, bound_mass=None):
    builder = newton.ModelBuilder()
    builder.validate_inertia_detailed = detailed
    if bound_mass is not None:
        builder.bound_mass = bound_mass

    body = builder.add_body(
        mass=mass,
        inertia=wp.mat33(inertia.astype(np.float32)),
        label="test_body",
    )

    model = builder.finalize()
    return (
        float(model.body_mass.numpy()[body]),
        np.array(model.body_inertia.numpy()[body]),
        builder.body_mass[body],
    )


print("bound-mass zero-mass case")
print(run_case(True, 0.0, np.diag([1.0, 1.0, 1.0]), bound_mass=0.1))
print(run_case(False, 0.0, np.diag([1.0, 1.0, 1.0]), bound_mass=0.1))

print("exact triangle boundary")
print(run_case(True, 1.0, np.diag([1.0, 1.0, 2.0])))
print(run_case(False, 1.0, np.diag([1.0, 1.0, 2.0])))

print("non-symmetric inertia")
print(
    run_case(
        True,
        1.0,
        np.array([[1.0, 2.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
    )
)
print(
    run_case(
        False,
        1.0,
        np.array([[1.0, 2.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]),
    )
)
```

## Expected Behavior

Validation should be semantically identical regardless of whether the user
enables detailed warnings.

It is fine for the two modes to differ in:

- warning granularity
- performance

It is not fine for them to differ in:

- corrected body mass
- corrected body inertia
- whether a tensor is considered valid
- whether invalid values are repaired vs left in place
- whether `finalize()` succeeds vs crashes
- whether `ModelBuilder` internal state is updated

## Proposed Direction

Use the detailed path as the semantic spec, then make the fast path match it,
with one important exception:

- first fix the `NaN` / `Inf` bug in `verify_and_correct_inertia()`

Recommended unification targets:

1. Decide and document the intended `bound_mass` behavior for zero and negative
   mass.
2. Enforce symmetry before eigen decomposition in both paths.
3. Guarantee strictly positive-definite inertia for positive-mass bodies in
   both paths, or explicitly decide not to do that and update both paths to
   agree.
4. Make exact triangle-inequality equality a no-op in both paths.
5. Handle `NaN` / `Inf` deterministically in both paths.
6. Make `ModelBuilder` post-finalize state consistent between modes.

## Suggested Tests

Add parity tests that compare the detailed and fast paths for:

- negative mass
- zero mass with non-zero inertia
- zero mass with `bound_mass`
- negative mass with `bound_mass`
- positive mass with zero inertia
- positive mass with semidefinite inertia
- non-symmetric inertia
- exact triangle-inequality boundary (`diag(1,1,2)`)
- near-boundary triangle-inequality cases
- `NaN` / `Inf` inertia

### PR DESCRIPTION
**Pull request by vreutskyy**
## Description

Unify the detailed (CPU) and fast (GPU kernel) inertia validation paths
in `ModelBuilder.finalize()` so both produce semantically identical results.

Fixes 7 inconsistencies documented in #2010:
1. `bound_mass` no longer overrides zero mass in GPU path (zero = static body)
2. GPU path now symmetrizes inertia before eigendecomposition
3. Both paths repair singular/near-zero inertia for positive-mass bodies
4. Triangle-inequality tolerance aligned to `1e-6` in both paths (float32)
5. NaN/Inf handled deterministically in both paths (zero out mass + inertia)
6. Fixed `UnboundLocalError` crash in CPU path on NaN eigenvalue input
7. `finalize()` is now side-effect-free on builder state — corrected values
   live only on the returned `Model`, consistent between both modes

Design decision for #7: rather than making both paths mutate the builder
(which caused a ~12% finalize benchmark regression from the GPU→CPU sync),
we make **neither** path mutate it. The builder holds user input; the Model
holds validated output. This makes `finalize()` idempotent and avoids
expensive device transfers.

Closes #2010

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (if user-facing change)

## Test plan

```
uv run --extra dev -m newton.tests -k test_inertia_validation
```

30 tests: parity tests comparing detailed vs fast path across NaN/Inf,
zero mass with bounds, singular inertia, non-symmetric tensors,
triangle-inequality boundary cases, plus a builder-state-unchanged test.

```
uv run --extra dev -m newton.tests -k test_inertia
```

Full inertia test suite passes with no regressions.

## Bug fix

**Steps to reproduce:**

1. Create a `ModelBuilder` with `bound_mass=0.1`
2. Add a body with `mass=0.0` (static body) and valid inertia
3. Finalize with `validate_inertia_detailed=False` (default)
4. Observe mass is incorrectly set to `0.1` instead of staying `0.0`

Or:

1. Create a `ModelBuilder` with `validate_inertia_detailed=True`
2. Add a body with `mass=1.0` and `inertia=diag(NaN, 1.0, 1.0)`
3. Call `finalize()` — crashes with `UnboundLocalError`

**Minimal reproduction:**

```python
import numpy as np
import newton
import warp as wp


def run_case(detailed: bool, mass: float, inertia: np.ndarray, bound_mass=None):
    builder = newton.ModelBuilder()
    builder.validate_inertia_detailed = detailed
    if bound_mass is not None:
        builder.bound_mass = bound_mass

    body = builder.add_body(
        mass=mass,
        inertia=wp.mat33(inertia.astype(np.float32)),
        label="test_body",
    )

    model = builder.finalize()
    return (
        float(model.body_mass.numpy()[body]),
        np.array(model.body_inertia.numpy()[body]),
    )


# Zero mass with bound_mass — should stay 0 in both paths
print(run_case(True, 0.0, np.diag([1.0, 1.0, 1.0]), bound_mass=0.1))
print(run_case(False, 0.0, np.diag([1.0, 1.0, 1.0]), bound_mass=0.1))

# NaN inertia — should not crash
print(run_case(True, 1.0, np.diag([np.nan, 1.0, 1.0])))
print(run_case(False, 1.0, np.diag([np.nan, 1.0, 1.0])))
```

### CODE REVIEW

File: newton/_src/geometry/inertia.py [hunk 1]
```diff
@@ -786,13 +796,31 @@ def validate_and_correct_inertia_kernel(
...
+        or not wp.isfinite(mass)
+        or wp.isnan(diag_sum)
+        or not wp.isfinite(diag_sum)
+        or wp.isnan(upper_sum)
+        or not wp.isfinite(upper_sum)
+        or wp.isnan(lower_sum)
+        or not wp.isfinite(lower_sum)
+    ):
+        mass = 0.0
+        inertia = wp.mat33(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/inertia.py:814]: nit: `wp.isfinite` already returns `False` for NaN, so the `wp.isnan` checks are redundant. Trimming them halves the branch count per body in the kernel:

```python
    if (
        not wp.isfinite(mass)
        or not wp.isfinite(diag_sum)
        or not wp.isfinite(upper_sum)
        or not wp.isfinite(lower_sum)
    ):
```

File: newton/_src/sim/builder.py [hunk 2]
```diff
@@ -9825,8 +9824,18 @@ def _to_wp_array(data, dtype, requires_grad):
...
                         )
 
-                    # Directly use the corrected arrays on the Model (avoids double allocation)
-                    # Note: This means the ModelBuilder's internal state is NOT updated for the fast path
+                        # Update builder state to match kernel output
+                        corrected_masses = body_mass_array.numpy()
+                        corrected_inertias = body_inertia_array.numpy()
+                        corrected_inv_masses = body_inv_mass_array.numpy()
+                        corrected_inv_inertias = body_inv_inertia_array.numpy()
+                        for i in range(len(self.body_mass)):
```
Comment (adenzler-nvidia) [inline newton/_src/sim/builder.py]: This updates every body in the builder even when only a handful were corrected. For large models you could skip unchanged bodies using the `correction_flags` array that's already available:

```python
flags = correction_flags.numpy()
for i in range(len(self.body_mass)):
    if flags[i]:
        self.body_mass[i] = float(corrected_masses[i])
        self.body_inertia[i] = self._coerce_mat33(corrected_inertias[i])
        self.body_inv_mass[i] = float(corrected_inv_masses[i])
        self.body_inv_inertia[i] = self._coerce_mat33(corrected_inv_inertias[i])
```

Not a correctness issue since finalization is one-shot, but worth considering for models with thousands of bodies.

File: newton/_src/geometry/inertia.py (hunk 1 above)
Comment (vreutskyy) [reply newton/_src/geometry/inertia.py:814]: Good catch — simplified in ad05ca1d. Now checks each of the 9 coefficients individually with `wp.isfinite` (which already covers NaN), no more `wp.isnan` or summing.

File: newton/_src/sim/builder.py (hunk 2 above)
Comment (vreutskyy) [reply newton/_src/sim/builder.py]: Done in ad05ca1d — now uses `correction_flags` to skip unchanged bodies:

```python
flags = correction_flags.numpy()
for i in range(len(self.body_mass)):
    if flags[i]:
        self.body_mass[i] = float(corrected_masses[i])
        self.body_inertia[i] = self._coerce_mat33(corrected_inertias[i])
        self.body_inv_mass[i] = float(corrected_inv_masses[i])
        self.body_inv_inertia[i] = self._coerce_mat33(corrected_inv_inertias[i])
```

File: newton/_src/geometry/inertia.py [hunk 3]
```diff
@@ -801,6 +828,19 @@ def validate_and_correct_inertia_kernel(
...
+            inertia[0, 0],
+            (inertia[0, 1] + inertia[1, 0]) * 0.5,
+            (inertia[0, 2] + inertia[2, 0]) * 0.5,
+            (inertia[0, 1] + inertia[1, 0]) * 0.5,
+            inertia[1, 1],
+            (inertia[1, 2] + inertia[2, 1]) * 0.5,
+            (inertia[0, 2] + inertia[2, 0]) * 0.5,
+            (inertia[1, 2] + inertia[2, 1]) * 0.5,
+            inertia[2, 2],
+        )
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/inertia.py:843]: Two symmetrization parity gaps remain:

1. **Conditional vs unconditional symmetrization.** The CPU path (line 664) only symmetrizes when `np.allclose(inertia, inertia.T)` fails, so mildly asymmetric tensors within the default tolerance (`atol=1e-8`) pass through unchanged. The GPU kernel unconditionally symmetrizes every positive-mass tensor here. For inputs where the asymmetry is small enough to pass `allclose` but nonzero, the two paths produce different results.

   Fix: either always symmetrize on CPU too (drop the `allclose` guard), or add an `allclose` check in the kernel before symmetrizing.

2. **`was_corrected` not set for symmetrization.** The kernel symmetrizes here but never marks `was_corrected = True` for it. When symmetrization is the only correction, `correction_flags` stays false and the builder sync loop (`if flags[i]`) skips that body — leaving the builder with the original unsymmetrized tensor while the model has the symmetrized one.

Comment (vreutskyy) [reply newton/_src/geometry/inertia.py:843]: Both fixed in aacb65dc:

1. CPU path now always symmetrizes (dropped the `allclose` guard). The warning still fires only when asymmetry is detected, but the result is always `(I + I^T) / 2` — matches the GPU kernel.

2. GPU kernel now checks `wp.ddot(inertia - sym, inertia - sym) > 0.0` after symmetrization and sets `was_corrected = True` when it changes the tensor, so `correction_flags` is accurate and the builder sync loop picks it up.

File: newton/_src/sim/builder.py (hunk 2 above)
Comment (eric-heiden) [reply newton/_src/sim/builder.py]: I think this iteration over all bodies is super slow, I will post a follow-up benchmark.

Comment (vreutskyy) [reply newton/_src/sim/builder.py]: Fixed — the loop now only iterates over corrected bodies via `np.nonzero(correction_flags)` instead of all bodies. For the common case (few corrections), this touches only a handful of entries. Also switched to scalar `wp.mat33()` constructor to avoid the row-vector warning path.

**Comment by eric-heiden** [comment]
On my tests the slowdown seems to be addressed now. There just appears to be some benchmark failure on the CI, can you check where this comes from?

| Case | Earlier `main` | Earlier PR | Current head |
| --- | ---: | ---: | ---: |
| Validation block corrected, 4000, `cuda:0` | 4.84 ms | 23.29 ms | 4.72 ms |
| Validation block corrected, 16000, `cuda:0` | 18.88 ms | 96.69 ms | 17.66 ms |
| Validation block corrected, 4000, `cpu` | 4.38 ms | 23.66 ms | 4.36 ms |
| `finalize()` corrected (reuse), 4000, `cuda:0` | 41.06 ms | 61.83 ms | 41.75 ms |
| `finalize()` corrected (reuse), 16000, `cuda:0` | 160.97 ms | 234.35 ms | 149.71 ms |
| `finalize()` corrected (reuse), 4000, `cpu` | 37.52 ms | 58.83 ms | 38.67 ms |
